compiler/query_structure.rs

```diff
@@ -31,124 +63,331 @@ impl ParametrisedQueryStructure {
             }
-            AnnotatedStage::Delete { .. }
-            | AnnotatedStage::Select(_)
-            | AnnotatedStage::Sort(_)
-            | AnnotatedStage::Offset(_)
-            | AnnotatedStage::Limit(_)
-            | AnnotatedStage::Require(_)
-            | AnnotatedStage::Distinct(_)
-            | AnnotatedStage::Reduce(_, _) => {}
+            AnnotatedStage::Update { block, annotations, .. } => {
+                debug_assert!(block.conjunction().nested_patterns().is_empty());
+                let block = self.add_block_impl(None, block.conjunction().constraints(), &annotations);
+                self.query_structure.stages.push(QueryStructureStage::Update { block });
+            }
+            AnnotatedStage::Delete { block, deleted_variables, annotations, .. } => {
+                debug_assert!(block.conjunction().nested_patterns().is_empty());
+                let block = self.add_block_impl(None, block.conjunction().constraints(), &annotations);
+                self.query_structure
+                    .stages
+                    .push(QueryStructureStage::Delete { block, deleted_variables: vec_from(deleted_variables.iter()) });
+            }
+            AnnotatedStage::Select(select) => {
+                self.query_structure.stages.push(QueryStructureStage::Select { variables: vec_from(select.variables.iter()) })
+            }
+            AnnotatedStage::Sort(sort) => {
+                self.query_structure.stages.push(QueryStructureStage::Sort { variables: vec_from(sort.variables.iter()) })
+            }
+            AnnotatedStage::Offset(offset) => {
+                self.query_structure.stages.push(QueryStructureStage::Offset { offset: offset.offset() })
+            }
+            AnnotatedStage::Limit(limit) => self.query_structure.stages.push(QueryStructureStage::Limit { limit: limit.limit() }),
+            AnnotatedStage::Require(require) => {
+                self.query_structure.stages.push(QueryStructureStage::Require { variables: vec_from(require.variables.iter()) })
+            }
+            AnnotatedStage::Distinct(_) => self.query_structure.stages.push(QueryStructureStage::Distinct),
+            AnnotatedStage::Reduce(reduce, _) => self.query_structure.stages.push(QueryStructureStage::Reduce {
+                reducers: vec_from(reduce.assigned_reductions.iter()),
+                groupby: vec_from(reduce.groupby.iter()),
+            }),
         }
-    });
-    Some(ParametrisedQueryStructure { branches, resolved_labels, calls_syntax: function_calls_syntax })
-}
+    }
 
-fn extract_query_structure_from_branch(
-    branches: &mut [Option<Vec<Constraint<Variable>>>; 64],
-    branch_id: BranchID,
-    conjunction: &Conjunction,
-) {
-    if branches[branch_id.0 as usize].is_none() {
-        branches[branch_id.0 as usize] = Some(Vec::new());
+    fn add_block(
+        &mut self,
+        existing_branch_id: Option<BranchID>,
```

Comment:
We always assume our existing branch IDs exist I thought? Since those are u16's, that's pretty safe

Reply:
They may not have been allocated a branch_id